rfcs/react-components/convergence/slot-children-render-function.md

```diff
@@ -0,0 +1,526 @@
+
+https://github.com/microsoft/fluentui/blob/220b321042b4bcc126c86a91811a34b04934d49b/packages/react-components/react-accordion/src/components/AccordionHeader/renderAccordionHeader.tsx#L13-L25
+
+Requirements for children render function:
+
+- should be able to override the entire slot
+- When rendering the original children, should respect nested slots.
+
+## Problem statement
+
+The main problem revolves around coalescing those 3 instances of properties a slot might have! At the moment,
+our current slot declaration mechanism revolves around 2 separate methods used in 2 different moments:
+
+1. `resolveShorthand`
+2. `getSlots`
+
+`resolveShorthand` is invoked at `useState` hook, it will receive a shorthand and will convert it to either `undefined` or an object comprising
+the merging of the External properties and Internal ones, following given priorities (`props` wins over `defaultProps`).
+
+`getSlots` is invoked at `render` method and will go over every single slot declared on `state.components` and convert every resolved shorthand into a pair of
+element to be render and properties to be passed. In this `getSlots` method, we also ensure that if a given resolved shorthand `children`
+property is a function, than we'll invoke that function:
+
+```ts
+// getSlots
+
+if (typeof children === 'function') {
+  const render = children;
+  // in this case the pair of element to be render and properties to be passed are converted into
+  // a Fragment and {children} where `children` will be the return of invoking the `children` render method
+  return [
+    React.Fragment,
+    {
+      children: render(slot, propsWithoutAsAndChildren),
+    } as unknown as R[K],
+  ];
+}
+```
+
+### Problem 1
+
+The `slot` argument will represent the merging between `state.component.button` value
+(back into the `AccordionHeader` `button` slot example) and `props.button.as` value. `propsWithoutAsAndChildren`
+as the name suggests is all the properties provided by the `resolveShorthand` invocation without including `as` and also `children`
+(since `children` by itself is the render function).
+
+Since we lost the Internal properties on the merging between it and the External ones
+There's no way to provide such the default declared `children` to the render function!
+
+### Problem 2
+
+At `render` method after `getSlots` invocation we have all `slots` and `slotProps` well defined. For the case of wrappers
+we'll also provide an override for `children`, taking precedence over External properties and also Internal properties.
+This will cause us to lose any children provided by a `render` function passed as a children.
```

Comment:
Yes. That's exactly the case for wrappers at the moment.

Reply:
````suggestion
This will cause us to lose any children provided by a `render` function passed as a children.

```js
function renderComponent(state) {
  const { slots, slotProps } = getSlots(state)

  return <slots.root {...slotProps.root}>Foo</slots.root>
}

const template = renderComponent({ root: { children: 'Baz' }})

template.props.children === 'Baz' // false
template.props.children === 'Foo' // true

// 💥 "Foo" wins over input passed by a user
```
````